### PR TITLE
feat(Warriorcats): add presence

### PR DESCRIPTION
**Pull request by PrincessAkira**
## Acknowledgements
- [X] I read the [Presence Guidelines](https://github.com/PreMiD/Presences/blob/main/.github/CONTRIBUTING.md)
- [X] I linted the code by running `yarn format`
- [X] The PR title follows the repo's [commit conventions](https://github.com/PreMiD/Presences/blob/main/.github/COMMIT_CONVENTION.md)

## Screenshots
<details>
<summary> Proof showing the creation/modification is working as expected </summary>

image

image

image

</details>

### CODE REVIEW

File: websites/W/Warriorcats/metadata.json [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+	"$schema": "https://schemas.premid.app/metadata/1.9",
+	"author": {
+		"id": "202740603790819328",
+		"name": "aimslut"
+	},
+	"service": "Warriorcats",
+	"description": {
+		"en": "meow"
+	},
+	"url": "warriorcats.de",
```
Comment (theusaf) [inline websites/W/Warriorcats/metadata.json:11]: The `www` subdomain is supported by the website, so it should be added here as well.
```suggestion
	"url": [
		"warriorcats.de",
		"www.warriorcats.de"
	],
```

File: websites/W/Warriorcats/metadata.json [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+	"version": "1.0.0",
+	"logo": "https://upload.wikimedia.org/wikipedia/commons/thumb/c/cf/Warrior_Cats_logo.svg/2560px-Warrior_Cats_logo.svg.png",
```
Comment (theusaf) [inline websites/W/Warriorcats/metadata.json:13]: The [logo](https://github.com/PreMiD/Presences/blob/main/.github/CONTRIBUTING.md#logo) should be a 512x512px **square**.

File: websites/W/Warriorcats/presence.ts [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+presence.on("UpdateData", async () => {
+	const presenceData: PresenceData = {
+		details: "Viewing an unsupported page",
+		state: "",
+		largeImageKey:
+			"https://upload.wikimedia.org/wikipedia/commons/thumb/c/cf/Warrior_Cats_logo.svg/2560px-Warrior_Cats_logo.svg.png",
+		startTimestamp: browsingTimestamp,
+		buttons: [
+			{ label: "Current Page", url: "https://www.warriorcats.com/" },
+			{ label: "Warrior Cats Germany", url: "https://www.warriorcats.de/" },
```
Comment (theusaf) [inline websites/W/Warriorcats/presence.ts:22]: [Buttons](https://github.com/PreMiD/Presences/blob/main/.github/CONTRIBUTING.md#presencets) that lead to a home page are now allowed.

File: websites/W/Warriorcats/presence.ts [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+	// find <article></article> and get the #post- number
+	let postNumber = document.querySelector("article").getAttribute("id");
+	postNumber = postNumber.replace(/post-/, "");
+	return postNumber;
+}
+
+presence.on("UpdateData", async () => {
+	const presenceData: PresenceData = {
+		details: "Viewing an unsupported page",
+		state: "",
```
Comment (theusaf) [inline websites/W/Warriorcats/presence.ts:16]: Empty state variable can be removed.
```suggestion
```

File: websites/W/Warriorcats/presence.ts [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+			{ label: "Warrior Cats Germany", url: "https://www.warriorcats.de/" },
+		],
+	};
+
+	switch (document.location.pathname) {
+		case "/": {
+			presenceData.details = "Viewing";
+			presenceData.state = "Warrior Cats Home";
+			presenceData.largeImageKey =
+				"https://www.warriorcats.de/wp-content/uploads/9783407759085-web.jpg";
```
Comment (theusaf) [inline websites/W/Warriorcats/presence.ts:31]: Assets like these should ideally be a square.

File: websites/W/Warriorcats/presence.ts [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+			presenceData.buttons = [
+				{ label: "Current Page", url: document.location.href },
```
Comment (theusaf) [inline websites/W/Warriorcats/presence.ts:33]: Rather than a generic "View Current Page" button, prefer using buttons with specific labels and which lead to useful content that a user may want to visit.

File: websites/W/Warriorcats/presence.ts [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+		buttons: [
+			{ label: "Current Page", url: "https://www.warriorcats.com/" },
+			{ label: "Warrior Cats Germany", url: "https://www.warriorcats.de/" },
+		],
+	};
+
+	switch (document.location.pathname) {
+		case "/": {
+			presenceData.details = "Viewing";
+			presenceData.state = "Warrior Cats Home";
```
Comment (theusaf) [inline websites/W/Warriorcats/presence.ts:29]: For pages like this, prefer putting everything into `details` and omitting `state`. (`Viewing Warrior Cats Home`, for example)

In addition, the text should support the site's [official language](https://github.com/PreMiD/Presences/blob/main/.github/CONTRIBUTING.md#presencets) (German).

File: websites/W/Warriorcats/presence.ts [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+			presenceData.details = "Viewing";
+			presenceData.buttons = [
+				{ label: "Current Page", url: document.location.href },
+				{ label: "Warrior Cats Germany", url: "https://www.warriorcats.de/" },
+			];
+			presenceData.state = "Warrior Cats Graphic Novels";
+			presenceData.largeImageKey =
+				"https://www.warriorcats.de/wp-content/uploads/9783407759085-web.jpg";
+			presence.setActivity(presenceData);
+
```
Comment (theusaf) [inline websites/W/Warriorcats/presence.ts:109]: Since you do this at the end of the function anyways, this is not needed.

File: websites/W/Warriorcats/presence.ts [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+		const staffelName = document.querySelector(
+				`#post-${postNumber} > div > div.entry-content > div > div.wc-biblio > div.wc-staffel`
+			).textContent,
+			// Remove "Warrior Cats |"
+
+			// remove Staffel
+
+			// get staffel number from url e.g
+
+			// for ressult3 match replacethis with result2 and replacce it with nothing
```
Comment (theusaf) [inline websites/W/Warriorcats/presence.ts:214]: These comments and others like these can be either removed or simplified. Some of them simply re-state what is obvious from the code.

File: websites/W/Warriorcats/presence.ts [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+			{ label: "Warrior Cats Germany", url: "https://www.warriorcats.de/" },
+		];
+	}
+
+	// https://www.warriorcats.de/beute/staffel-viii-band-2-wolken/
+	// extract Staffel number and book number and book name from URL
+	// like this Staffel 8 Band 2 Wolken
+	// and then set the presence state to Currently Staffel 8 Band 2 Wolken
+
+	async function getStaffelName(postNumber: string) {
```
Comment (theusaf) [inline websites/W/Warriorcats/presence.ts:204]: These functions do not do anything `async`. Also, for better performance/readability, these functions should be defined outside of the `UpdateData` loop.

File: websites/W/Warriorcats/presence.ts [hunk 11]
```diff
@@ -0,0 +1,270 @@
+const presence = new Presence({
+		clientId: "936630095035113522",
+	}),
+	browsingTimestamp = Math.floor(Date.now() / 1000);
+
+async function getPostNumber() {
+	// find <article></article> and get the #post- number
+	let postNumber = document.querySelector("article").getAttribute("id");
+	postNumber = postNumber.replace(/post-/, "");
+	return postNumber;
```
Comment (theusaf) [inline websites/W/Warriorcats/presence.ts:10]: This code (and other functions) can be simplified to a single `return` statement.

File: websites/W/Warriorcats/presence.ts [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+
+		presenceData.state = `Viewing ${buchName}`;
+		presenceData.details = `${staffelName}`;
+		presenceData.largeImageKey = buchImage;
+		presenceData.buttons = [
+			{ label: "Current Page", url: document.location.href },
+			{ label: "Warrior Cats Germany", url: "https://www.warriorcats.de/" },
+		];
+	}
+	presence.clearActivity();
```
Comment (theusaf) [inline websites/W/Warriorcats/presence.ts:268]: Since the activity is set immediately afterwards, this is not needed, and could cause problems.
```suggestion
```

File: websites/W/Warriorcats/metadata.json [hunk 13]
```diff
@@ -0,0 +1,20 @@
+{
+	"$schema": "https://schemas.premid.app/metadata/1.9",
+	"author": {
+		"id": "202740603790819328",
+		"name": "aimslut"
+	},
+	"service": "Warriorcats",
+	"description": {
+		"en": "meow"
```
Comment (theusaf) [inline websites/W/Warriorcats/metadata.json:9]: This is not the official description of the website. Add the official description and/or translate if needed.

File: websites/W/Warriorcats/metadata.json [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+	},
+	"url": "warriorcats.de",
+	"version": "1.0.0",
+	"logo": "https://upload.wikimedia.org/wikipedia/commons/thumb/c/cf/Warrior_Cats_logo.svg/2560px-Warrior_Cats_logo.svg.png",
+	"thumbnail": "https://upload.wikimedia.org/wikipedia/commons/thumb/c/cf/Warrior_Cats_logo.svg/2560px-Warrior_Cats_logo.svg.png",
+	"color": "#ffffff",
+	"category": "other",
+	"tags": [
+		"anime"
+	]
```
Comment (theusaf) [inline websites/W/Warriorcats/metadata.json:19]: This website does not appear to be related to anime.

**Review by theusaf — CHANGES_REQUESTED** [review]

**Comment by PrincessAkira** [comment]
I misspressed.

File: websites/W/Warriorcats/metadata.json (hunk 13 above)
Comment (PrincessAkira) [reply websites/W/Warriorcats/metadata.json:9]: Oh yeah. That was just placeholer. I forgot that :D

File: websites/W/Warriorcats/presence.ts [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+		.getAttribute("id")
+		.replace(/post-/, "");
+}
+
+presence.on("UpdateData", async () => {
+	const presenceData: PresenceData = {
+		details: "Stöbert auf der Warrior Cats Seite",
+		largeImageKey: "https://i.imgur.com/Ij2BuGr.jpg",
+		startTimestamp: browsingTimestamp,
+		buttons: [{ label: "Current Page", url: document.location.href }],
```
Comment (theusaf) [inline websites/W/Warriorcats/presence.ts:18]: This [button](https://github.com/PreMiD/Presences/blob/main/.github/CONTRIBUTING.md#presencets) could lead to a main page and is also very generic.
Rather than deleting buttons on paths that shouldn't have buttons, add specific buttons instead.

File: websites/W/Warriorcats/presence.ts [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+			break;
+		}
+		case "/beutehaufen-uebersicht/": {
+			presenceData.details = "Beutehaufen Übersicht";
+			break;
+		}
+		case "/nachrichten-vom-baumgeviert/": {
+			presenceData.details = "lesen der Nachrichten";
+
+			break;
```
Comment (theusaf) [inline websites/W/Warriorcats/presence.ts:34]: Ensure to remove the extra line before the breaks

File: websites/W/Warriorcats/presence.ts [hunk 17]
```diff
@@ -0,0 +1,134 @@
...
+	}
+
+	async function getBuchImage(postNumber: string) {
+		const buchImage = document
+			.querySelector(
+				`#post-${postNumber} > div > div.entry-content > div > div.wc-article-thumbnail > a > img`
+			)
+			.getAttribute("src");
+		return buchImage;
+	}
```
Comment (theusaf) [inline websites/W/Warriorcats/presence.ts:118]: Ditto for these (simplify one-use variables, remove unused `async`)

File: websites/W/Warriorcats/presence.ts [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+		document.location.pathname.startsWith("/beute/staffel-") ||
+		document.location.pathname.startsWith("/beute/")
+	) {
+		const postNumber = await getPostNumber(),
+			staffelName = await getStaffelName(postNumber),
+			buchName = await getBuchName(postNumber),
+			buchImage = await getBuchImage(postNumber);
+
+		presenceData.state = `Viewing ${buchName}`;
+		presenceData.details = `${staffelName}`;
```
Comment (theusaf) [inline websites/W/Warriorcats/presence.ts:130]: Since the function already returns a string, this may not be needed? Also, these may be one-use variables that can be inlined.
```suggestion
		presenceData.details = staffelName;
```

**Review by theusaf — CHANGES_REQUESTED** [review]

File: websites/W/Warriorcats/presence.ts [hunk 19]
```diff
@@ -0,0 +1,134 @@
...
+		default:
+			{
+				presenceData.details = "unbekannte Seite";
+			}
+
+			if (document.location.pathname.startsWith("/staffel-")) {
+				presenceData.details = `Staffel ${
+					document.location.pathname.split("/")[1].split("-")[1]
+				}`;
+			}
```
Comment (theusaf) [inline websites/W/Warriorcats/presence.ts:85]: Place the if statement within the block, or use an else statement

File: websites/W/Warriorcats/presence.ts [hunk 20]
```diff
@@ -0,0 +1,134 @@
...
+			delete presenceData.buttons;
+			presenceData.details = "Auf der Startseite";
+			break;
+		}
+		case "/beutehaufen-uebersicht/": {
+			presenceData.details = "Beutehaufen Übersicht";
+			break;
+		}
+		case "/nachrichten-vom-baumgeviert/": {
+			presenceData.details = "lesen der Nachrichten";
```
Comment (theusaf) [inline websites/W/Warriorcats/presence.ts:32]: Make sure that the texts are consistently capitalized.

**Review by theusaf — CHANGES_REQUESTED** [review]

File: websites/W/Warriorcats/presence.ts [hunk 21]
```diff
@@ -0,0 +1,138 @@
...
+		.getAttribute("id")
+		.replace(/post-/, "");
+}
+
+presence.on("UpdateData", async () => {
+	const presenceData: PresenceData = {
+		details: "Stöbert auf der Warrior Cats Seite",
+		largeImageKey: "https://i.imgur.com/Ij2BuGr.jpg",
+		startTimestamp: browsingTimestamp,
+	};
```
Comment (DarkVillager) [inline websites/W/Warriorcats/presence.ts:18]:
```suggestion
	},
	{href, pathname} = document.location;
```

**Review by DarkVillager — CHANGES_REQUESTED** [review]

File: websites/W/Warriorcats/presence.ts [hunk 22]
```diff
@@ -0,0 +1,138 @@
...
+		presenceData.buttons = [
+			{
+				label: "Staffel ansehen",
+				url: document.location.href,
+			},
+		];
+		presenceData.details = `Staffel ${
+			document.location.pathname.split("/")[1].split("-")[1]
+		}`;
+	}
```
Comment (DarkVillager) [inline websites/W/Warriorcats/presence.ts:105]:
```suggestion
	switch (pathname) {
		case "/": {
			presenceData.details = "auf der Startseite";
			break;
		}
		case "/beutehaufen-uebersicht/": {
			presenceData.details = "Beutehaufen Übersicht";
			break;
		}
		case "/nachrichten-vom-baumgeviert/": {
			presenceData.details = "lesen der Nachrichten";
			break;
		}
		case "/nutzerkonto/": {
			presenceData.details = "im Nutzerkonto";
			break;
		}
		case "/dein-clanlager/": {
			presenceData.details = "im  Clan Camp";
			break;
		}
		case "/deine-clankatze/": {
			presenceData.details = "mit der eigenen Katze";
			break;
		}
		case "/graphic-novel/": {
			presenceData.details = "in den Graphic Novels";
			presenceData.buttons = [
				{
					label: "Buch ansehen",
					url: href,
				},
			];
			break;
		}
		case "/alle-baende/": {
			presenceData.details = "in der Liste aller Bände";
			break;
		}
		case "/erin-hunter/": {
			presenceData.details = "auf der Erin Hunter Seite";
			break;
		}
		case "/welt-der-warrior-cats/": {
			presenceData.details = "betrachtet die Welt";
			break;
		}
		case "/zeitstrahl/": {
			presenceData.details = "die Lore der Warrior Cats";
			break;
		}
		case "/wiki/": {
			presenceData.details = "im Wiki lesen";
			break;
		}
		default: {
			presenceData.details = "unbekannte Seite";
		}
	}

	if (
		pathname.startsWith("/beute/staffel-") ||
		pathname.startsWith("/beute/")
	) {
		const postNumber = getPostNumber();

		presenceData.state = `Viewing ${getBuchName(postNumber)}`;
		presenceData.details = getStaffelName(postNumber);
		presenceData.buttons = [
			{
				label: "Buch ansehen",
				url: href,
			},
		];
		presenceData.largeImageKey = getBuchImage(postNumber);
	} else if (pathname.startsWith("/staffel-")) {
		presenceData.buttons = [
			{
				label: "Staffel ansehen",
				url: href,
			},
		];
		presenceData.details = `Staffel ${
			pathname.split("/")[1].split("-")[1]
		}`;
	}
```

**Review by DarkVillager — CHANGES_REQUESTED** [review]

**Review by theusaf — APPROVED** [review]

**Review by Bas950 — APPROVED** [review]

**Comment by NULLYUKI** [comment]
WarriorCats also provides an English website which is reachable under this domain https://www.warriorcats.com/
I feel like this would be a welcome change to include in the metadata.json file?
